Implement new MemoryDB and make DB selectable at runtime

halfalicious opened the pull request:
These changes take care of the first 2 items in #5262 - renaming the "old" MemoryDB to StateCacheDB and creating a new thread-safe MemoryDB implementation which implements `DatabaseFace `and is backed by `std::unordered_map`. I've verified that the changes compile clean in both the Debug and RelWithDebInfo configurations and that the StateCacheDB tests all still pass: 

```
C:\Users\nilse\Documents\Code\aleth\build\test\Debug>testeth -t stateCacheDB
Running tests using path: "../../test/jsontests"
Running 7 test cases...
Test Case "kill":
Test Case "purgeMainMem":
Test Case "purgeMainMem_Refs":
Test Case "purgeAuxMem":
Test Case "copy":
Test Case "lookUp":
Test Case "stream":

*** No errors detected

C:\Users\nilse\Documents\Code\aleth\build\test\Debug>
```

gumb0 commented on libdevcore/MemoryDB.cpp:
```diff
@@ -1,189 +1,104 @@
+    }
+    MemoryDBBatch batch = batchPtr->writeBatch();
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (const auto& e : batch)
```
we put `const` after type specifier, so this should be `auto const&`

gumb0 commented on libdevcore/MemoryDB.cpp:
```diff
@@ -1,189 +1,104 @@
-            ret.insert(i.first);
-    return ret;
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (const auto& e : m_db)
```
`auto const&`

gumb0 commented:
Note that we usually rebase PR branch on the `master` instead of merging `master` into branch, to keep history more linear.

chfast commented:
I have not reviewed it all yet, but will be easier to handle the review if the first two commits are sent as a separate PR.

halfalicious commented:
> Note that we usually rebase PR branch on the `master` instead of merging `master` into branch, to keep history more linear.

Makes sense, I'll look into how to do that after the PR is signed off.

halfalicious commented on libdevcore/MemoryDB.h:
```diff
@@ -14,79 +14,37 @@
 private:
-    MemoryDB const& m_o;
-    bool m_r;
+        std::unordered_map<std::string, std::string> m_db;
```
DatabaseFace::lookup returns std::string, I'll look into how LevelDB goes from slices to strings before returning the data.

halfalicious commented on libdevcore/MemoryDB.cpp:
```diff
@@ -1,189 +1,104 @@
-    if (it != m_main.end())
+    std::string value;
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_db.count(_key.data()))
```
Find can return empty so I still need to check for that before returning. Doing that check is probably cheaper than calling count() though so I'll go with that approach instead. Thanks for calling this out!

gumb0 commented:
Last commit is not linked to your account, I think you can fix it by adding an email you used (in git) to your github profile

gumb0 commented on libdevcore/MemoryDB.h:
```diff
@@ -40,11 +40,11 @@ class MemoryDB : public DatabaseFace
-        std::unordered_map<std::string, std::string> m_db;
-        mutable std::mutex m_mutex;
+    std::unordered_map<std::string, std::string> m_db;
+    mutable dev::Mutex m_mutex;
```
`dev::` is not really needed, we're in `dev`

gumb0 commented on libdevcore/MemoryDB.cpp:
```diff
@@ -33,30 +49,31 @@ void MemoryDBWriteBatch::kill(Slice _key)
 {
-    std::lock_guard<std::mutex> lock(m_mutex);
+    dev::Guard lock(m_mutex);
     return m_db.count(_key.data()) != 0;
```
must be `_key.toString()`

gumb0 replied:
Also in `MemoryDBWriteBatch` methods above

halfalicious replied:
Sorry about that, I thought I had gotten all of them.

gumb0 commented on libdevcore/MemoryDB.cpp:
```diff
@@ -33,30 +49,31 @@ void MemoryDBWriteBatch::kill(Slice _key)
 bool MemoryDB::exists(Slice _key) const
 {
-    std::lock_guard<std::mutex> lock(m_mutex);
+    dev::Guard lock(m_mutex);
```
can be just `Guard`

gumb0 commented on libdevcore/MemoryDB.cpp:
```diff
@@ -88,17 +106,17 @@ void MemoryDB::commit(std::unique_ptr<WriteBatchFace> _batch)
+    for (auto const& e : m_db)
     {
-        if (!f(Slice(e.first.c_str()), Slice(e.second.c_str())))
+        if (!_f(Slice(e.first.c_str()), Slice(e.second.c_str())))
```
Constructing `Slice` from `c_str()` also assumes it's zero-terminated string, which it is not. Look at other `vector_ref` constructors, is there constructor just from `std::string` ?

halfalicious commented:
> Last commit is not linked to your account, I think you can fix it by adding an email you used (in git) to your github profile

Thanks for the heads up I didn't even notice that, I've fixed it.

chfast commented:
Also, this should be rebased to remove the "Merge" commit.

halfalicious commented:
I've fixed the issues that @gumb0  called out and have created a set of unit tests which cover everything but `foreach `(I'm planning on creating the `forEach `tests tomorrow). Once I've created the foreach tests and verified they all pass I'll push my latest changes (I'm hoping to do this tomorrow evening) and resume validating my `DBFactory `changes. Some notes/questions:

- I added a size function to both `MemoryDB `and `MemoryDBWriteBatch `for testing purposes (so the unit tests can validate how many elements have been inserted into the db / write batch, respectively). I chose this approach over making the MemoryDB test suite a `friend `class since it seems cleaner plus the functions can potentially have utility outside of testing. Thoughts? 
- I moved the `MemoryDBWriteBatch `class declaration from the CPP file to the header file so that the unit tests have access to it so they can validate it. Let me know if you'd like me to go with a different approach. 
- I'm running into build issues again on my home machine, even after completely uninstalling Visual Studio 2017. I'll create a new issue for tracking purposes and try to untangle it tomorrow / this weekend.

gumb0 commented:
The testing approach is ok I think.

halfalicious commented on libdevcore/MemoryDB.cpp:
```diff
@@ -14,176 +14,109 @@
     {
-        it->second.first = _v.toString();
-        it->second.second++;
+        value = it->first;
```
This should be it->second

halfalicious commented:
> I've fixed the issues that @gumb0 called out and have created a set of unit tests which cover everything but `foreach`(I'm planning on creating the `forEach`tests tomorrow). Once I've created the foreach tests and verified they all pass I'll push my latest changes (I'm hoping to do this tomorrow evening) and resume validating my `DBFactory`changes. Some notes/questions:
> 
> * I added a size function to both `MemoryDB`and `MemoryDBWriteBatch`for testing purposes (so the unit tests can validate how many elements have been inserted into the db / write batch, respectively). I chose this approach over making the MemoryDB test suite a `friend`class since it seems cleaner plus the functions can potentially have utility outside of testing. Thoughts?
> * I moved the `MemoryDBWriteBatch`class declaration from the CPP file to the header file so that the unit tests have access to it so they can validate it. Let me know if you'd like me to go with a different approach.
> * I'm running into build issues again on my home machine, even after completely uninstalling Visual Studio 2017. I'll create a new issue for tracking purposes and try to untangle it tomorrow / this weekend.

U

> I've fixed the issues that @gumb0 called out and have created a set of unit tests which cover everything but `foreach`(I'm planning on creating the `forEach`tests tomorrow). Once I've created the foreach tests and verified they all pass I'll push my latest changes (I'm hoping to do this tomorrow evening) and resume validating my `DBFactory`changes. Some notes/questions:
> 
> * I added a size function to both `MemoryDB`and `MemoryDBWriteBatch`for testing purposes (so the unit tests can validate how many elements have been inserted into the db / write batch, respectively). I chose this approach over making the MemoryDB test suite a `friend`class since it seems cleaner plus the functions can potentially have utility outside of testing. Thoughts?
> * I moved the `MemoryDBWriteBatch`class declaration from the CPP file to the header file so that the unit tests have access to it so they can validate it. Let me know if you'd like me to go with a different approach.
> * I'm running into build issues again on my home machine, even after completely uninstalling Visual Studio 2017. I'll create a new issue for tracking purposes and try to untangle it tomorrow / this weekend.

I've added the foreach tests and verified that everything builds clean and that all tests pass locally. I'll revisit the DBFactory integration over the weekend.

halfalicious commented:
@chfast :Thanks for the tip about commit messages, makes sense (so one can easily tell by glancing at the commit what it covers), I'll be sure to do that in the future.

halfalicious commented:
My latest set of commits complete this issue (barring any bugs that need to be fixed before rebasing). The `DBFactory ` has been integrated and one can now select the database (LevelDB, which is the default selection, or the new MemoryDB) at runtime via the new '--db' parameter. I haven't performed any qualitative perf benchmarks but anecdotally, I kicked off a new sync using MemoryDB and blocks started being imported nearly immediately and were imported faster than in the LevelDB case. Some additional notes:
* Here's the new command-line output for the database options:
```

DATABASE OPTIONS:
  --db <name> (=leveldb)                                      Select database implementation. Available options are: leveldb, memorydb.
  --db-path <path> (=C:\Users\nilse\AppData\Roaming\Ethereum) Database path (for non-memory database options)
```

* @gumb0: I know that you mentioned that we don't need to log all database options, but I decided to do so because I think it improves usability (one can easily glance at the help output and know what the options are without having to go digging for documentation or looking through the code), the list isn't very long, and we list the various VM types for the VM option. If you still feel like we shouldn't list the DB types let me know and I can pull them out. 

* I've repurposed `--db-path` to specify the database path and created a new parameter (`--data-dir`) for other files (which, as far as I know, are the keystore, secret, and configuration files). 
* Given that `DBFactory `now includes the database path, I've pulled out the `Defaults` class. 
* Both the DB path and the "other files" path default to the same path
* I've created what I think to be a pretty exhaustive set of tests for the new MemoryDB. They are pretty straightforward and all pass.
* I updated the OverlayDB tests to explicitly specify LevelDB when creating the new DB via `DBFactory` despite it already being the default option in case we want to change the default in the future, and so it's clear to code readers which DB is being created. I've verified that the OverlayDB tests all still pass as well. 
* I pass the DB path to the `DBFactory::create` function, even when the MemoryDB option is set (the path is silently ignored), because I think this is cleaner than having a MemoryDB-specific special case for each DB creation call where a path isn't supplied.

chfast commented on aleth/main.cpp:
```diff
@@ -769,7 +770,7 @@ int main(int argc, char** argv)
         AccountManager::streamWalletHelp(cout);
         cout << clientDefaultMode << clientTransacting << clientNetworking << clientMining << minerOptions;
-        cout << importExportMode << vmOptions << loggingProgramOptions << generalOptions;
+        cout << importExportMode << vmOptions << dbOptions << loggingProgramOptions << generalOptions;
```
I think "db options" should be before "vm options".

chfast commented on libdevcore/DBFactory.cpp:
```diff
@@ -0,0 +1,151 @@
+        break;
+    default:
+        assert(false);
+        return std::unique_ptr<DatabaseFace>(new LevelDB(_path));
```
Return null unique_ptr. Can be simple as `return {}`.

chfast commented on libdevcore/DBFactory.cpp:
```diff
@@ -0,0 +1,151 @@
+    {
+    case DatabaseKind::LevelDB:
+        return std::unique_ptr<DatabaseFace>(new LevelDB(_path));
+        break;
```
`break`s after `return`s are redundant.

halfalicious replied:
Yes, thought I needed this to satisfy the compiler, since I don't I'll remove it. Thanks for calling this out!

chfast commented on libdevcore/MemoryDB.cpp:
```diff
@@ -14,176 +14,95 @@
-    if (it != m_main.end())
+    std::string value;
+    Guard lock(m_mutex);
+    auto const& it = m_db.find(_key.toString());
```
This can be simplified a bit as
```
if (it != m_db.end())
    return it->second;
return {};
```

chfast replied:
Also, for single-liners after `if` we skip `{}`.

halfalicious replied:
Thanks! Was aware of the skipping {} for single-line ifs but forgot in this case, thanks for catching this!

chfast commented on libdevcore/MemoryDB.cpp:
```diff
@@ -14,176 +14,95 @@
-    WriteGuard l(x_this);
-#endif
-    m_aux[_h] = make_pair(_v.toBytes(), true);
+    return std::unique_ptr<WriteBatchFace>(new MemoryDBWriteBatch());
```
Watch out with creating objects as `Object()`, compiler sometimes thinks it is a function definition (not in this context however). Always safe alternatives: `new Object`, `new Object{}`.

halfalicious replied:
Ah interesting, thanks for calling this out!

chfast commented on libdevcore/MemoryDB.cpp:
```diff
@@ -14,176 +14,95 @@
-            it = m_aux.erase(it);
+    if (!_batch)
+    {
+        BOOST_THROW_EXCEPTION(DatabaseError() << errinfo_comment("Cannot commit null batch"));
```
Can this happen?

gumb0 replied:
In practice it should, but can if someone calls `db->commit({});`

chfast replied:
Both exceptions here are rather candidates for asserts.

gumb0 commented on aleth/main.cpp:
```diff
@@ -394,12 +393,13 @@ int main(int argc, char** argv)
     auto addGeneralOption = generalOptions.add_options();
-    addGeneralOption("db-path,d", po::value<string>()->value_name("<path>"),
+    addGeneralOption("data-dir,d", po::value<string>()->value_name("<path>"),
         ("Load database from path (default: " + getDataDir().string() + ")").c_str());
```
Option description should be changed (Maybe `Load config files and keystore from path`)

halfalicious replied:
Ah yes, thank you for catching that!

gumb0 commented on libdevcore/DBFactory.cpp:
```diff
@@ -0,0 +1,151 @@
+auto g_kind = DatabaseKind::LevelDB;
+fs::path g_dbPath = getDataDir();
+
+/// A helper type to build the tabled of DB implementations.
```
the table

gumb0 commented on libdevcore/DBFactory.cpp:
```diff
@@ -0,0 +1,151 @@
+struct DBKindTableEntry
+{
+    DatabaseKind kind;
+    const char* name;
```
`char const*`

gumb0 commented on libdevcore/DBFactory.cpp:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    BOOST_THROW_EXCEPTION(std::system_error(std::make_error_code(std::errc::invalid_argument),
```
I see this comes from `setVMKind`, but I don't think it's a good idea to throw `system_error` here which is usually thrown from standard library classes working with the OS functions.

In general in C++ it's usually advised to define custom exception types deriving `std::exception`, we also utilize `boost::exception`, and the usual approach is to define in some relevant header exception type like
```
DEV_SIMPLE_EXCEPTION(InvalidDatabaseKind);
```

Then throw it here like
```
BOOST_THROW_EXCEPTION(InvalidDatabaseKind());
```
or if you want to include comment
```
BOOST_THROW_EXCEPTION(InvalidDatabaseKind() << errinfo_comment("invalid database kind supplied: " + _name));
```

gumb0 commented on libdevcore/DBFactory.h:
```diff
@@ -0,0 +1,60 @@
+    unsigned _lineLength = boost::program_options::options_description::m_default_line_length);
+
+bool isMemoryDB();
+DatabaseKind getDatabaseKind();
```
Getters are usually without `get` prefix, so `databaseKind()`

gumb0 commented on libdevcore/DBFactory.h:
```diff
@@ -0,0 +1,60 @@
+
+bool isMemoryDB();
+DatabaseKind getDatabaseKind();
+boost::filesystem::path getDatabasePath();
```
and `databasePath()`

gumb0 commented on libdevcore/DBFactory.cpp:
```diff
@@ -0,0 +1,151 @@
+        "invalid database kind supplied: " + _name));
+}
+
+void setDatabasePath(const std::string& _path)
```
`std::string const&`

gumb0 commented on libdevcore/MemoryDB.cpp:
```diff
@@ -14,176 +14,95 @@
+        BOOST_THROW_EXCEPTION(
+            DatabaseError() << errinfo_comment("Invalid batch type passed to MemoryDB::commit"));
+    }
+    std::unordered_map<std::string, std::string> batch = batchPtr->writeBatch();
```
This is unnecessary copy of the batch. If you wanted to simplify for loop below, you could just define this as const reference.

halfalicious replied:
Nice catch!

gumb0 commented:
> Can we also consider improving consistency of naming convention? I.e. "db" vs "database" name.

I'd rename `dmProgramOptions` to `databaseProgramOptions`

gumb0 commented:
> @gumb0: I know that you mentioned that we don't need to log all database options, but I decided to do so because I think it improves usability (one can easily glance at the help output and know what the options are without having to go digging for documentation or looking through the code), the list isn't very long, and we list the various VM types for the VM option. If you still feel like we shouldn't list the DB types let me know and I can pull them out.

My concern is that when we finish intergating RocksDB support, I think we shouldn't build and link both LevelDB and RocksDB at the same time, in other words it should be a compile-time option. So this way `--db` will always have just two possible values.
We can leave your variant now as it is, but I expect some code around setting DB kind might go away with RocksDB integration.

chfast commented:
> My concern is that when we finish intergating RocksDB support, I think we shouldn't build and link both LevelDB and RocksDB at the same time, in other words it should be a compile-time option. So this way `--db` will always have just two possible values.
> We can leave your variant now as it is, but I expect some code around setting DB kind might go away with RocksDB integration.

Building both is to so bad idea (not by default of curse) for easier performance comparison.

chfast commented:
For me it is ok to have only 2 options in the final binary at this point.

chfast commented:
I can rebase this PR when ready.

Why is AppVeyor slow again?

gumb0 commented on libethereum/BlockChain.cpp:
```diff
@@ -212,58 +211,70 @@ void BlockChain::init(ChainParams const& _p)
-                "already open. You appear to have another instance of ethereum running. Bailing.";
-            BOOST_THROW_EXCEPTION(DatabaseAlreadyOpen());
+            cwarn << "Unknown database error occurred during in-memory database creation";
+            BOOST_THROW_EXCEPTION(UnknownDatabaseError());
```
I would `throw;` here, current exception `ex` could contain some valuable info from lower level, it's better to make it visible than to suppress it.

gumb0 commented on libethereum/BlockChain.cpp:
```diff
@@ -343,10 +354,10 @@ void BlockChain::rebuild(fs::path const& _path, std::function<void(unsigned, uns
-    std::unique_ptr<db::DatabaseFace> oldExtrasDB(
-        new db::DBImpl(extrasPath / fs::path("extras.old")));
-    m_extrasDB.reset(new db::DBImpl(extrasPath / fs::path("extras")));
+    if (!db::isMemoryDB())
```
I think this `rebuild()` method doesn't make any sense for MemoryDB (it rebuilds the state DB by replaying all blocks from blockchain DB), so let's just return immediately in the beginning if it's MemoryDB (maybe output some log message, too)

halfalicious replied:
I took a look at when rebuild is called and it's only called when re-opening an existing database, which obviously doesn't make sense when running MemoryDB. I'll go ahead and return immediately.

gumb0 commented on libethereum/State.cpp:
```diff
@@ -60,40 +59,48 @@ State::State(State const& _s):
+        else
+        {
+            cwarn << "Unknown error when creating in-memory database.";
+            BOOST_THROW_EXCEPTION(UnknownDatabaseError());
```
Here too, I would instead do `if (db::isMemoryDB()) throw;` in the beginning of `catch`

gumb0 commented on test/unittests/libweb3core/overlaydb.cpp:
```diff
@@ -20,22 +20,23 @@
 {
     TransientDirectory td;
-    std::unique_ptr<db::DBImpl> db(new db::DBImpl(td.path()));
+	std::unique_ptr<db::DatabaseFace> db = DBFactory::create(DatabaseKind::LevelDB, td.path());
```
In all the tests in this file, I'd like to use `MemoryDB` instead of `LevelDB` and not use `TransientDirectory`, but this may go into separate PR.

halfalicious replied:
Got it, I'll take care of that in a separate PR

halfalicious replied:
Should we just create a copy of this file which uses MemoryDB since we presumably want test coverage for both the LevelDB and MemoryDB implementations (since LevelDB is the database actually used by users, at least until I get RocksDB online)?

gumb0 replied:
I don't think we need this. I believe unit tests by their nature should be as isolated as possible, so `OverlayDB` tests should test only the code in `OverlayDB`
(this is not quite the case here, since `MemoryDB` is used, so it's not ideal)

We could have additionally some set of tests checking how our `LevelDBDatabase` implementation works, but then `OverlayDB` shouldn't interfere with that.

chfast replied:
@gumb0 can you propose a code change and push it to this PR?

gumb0 replied:
I meant that we don't need additional test file as Nils proposed above, this code in this PR is fine

gumb0 commented:
> Why isn't it a good idea to build with both if we only use one at runtime, is it due to the increased binary size?

binary size and build time

gumb0 commented:
> Why is AppVeyor slow again?


I don't know, I thought it's the problem of timeout going down from 2 hours to 1 hour sometimes for unknown reason?

halfalicious commented on libdevcore/DBFactory.h:
```diff
@@ -0,0 +1,60 @@
+    ~DBFactory() = delete;
+
+    static std::unique_ptr<DatabaseFace> create();
+    static std::unique_ptr<DatabaseFace> create(boost::filesystem::path _path);
```
Makes sense!

halfalicious commented on libdevcore/MemoryDB.h:
```diff
@@ -14,79 +14,53 @@
 
-    h256Hash keys() const;
+    std::unordered_map<std::string, std::string> const& writeBatch() const { return m_batch; }
+    std::unordered_map<std::string, std::string>& writeBatch() { return m_batch; }
```
Correct, I don't use it anywhere. I'll remove it.

chfast commented:
The blockchain tests are 6x faster with MemoryDB!
```sh
# LevelDB:
test/testeth -t BlockchainTests  76,08s user 18,78s system 23% cpu 6:39,06 total
# MemoryDB:
test/testeth -t BlockchainTests  68,12s user 10,07s system 107% cpu 1:13,04 total
```

halfalicious commented:
> I did the rebase with some squashes. It's not perfect, but making it perfect would require splitting some commits (I don't know how to do it easily with preserving the author).

Awesome, thanks! I can take care of that next time :)
